drivers/adc/adc_emul.c

```diff
@@ -0,0 +1,512 @@
+/**
+ * @file
+ *
+ * @brief Emulated ADC driver
+ */
+
+/*
+ * Copyright 2021 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_adc_emul
+
+#include <drivers/adc.h>
+#include <drivers/adc/adc_emul.h>
+#include <kernel.h>
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+#include <zephyr.h>
+
+LOG_MODULE_REGISTER(adc_emul, CONFIG_ADC_LOG_LEVEL);
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define ADC_EMUL_MAX_RESOLUTION 16
+
+typedef uint16_t adc_emul_res_t;
+
+struct adc_emul_channel_cfg {
+	adc_emul_value_function func;
+	void *param;
+	enum adc_gain gain;
+	enum adc_reference ref;
+};
+
+struct adc_emul_config {
+	uint8_t channels_num;
+};
+
+struct adc_emul_data {
+	struct adc_context ctx;
+	const struct device *dev;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+	uint32_t channels;
+	uint16_t res_mask;
+	uint16_t ref_vdd;
+	uint16_t ref_ext0;
+	uint16_t ref_ext1;
+	uint16_t ref_int;
+	uint8_t oversampling;
+	uint8_t differential;
+	struct adc_emul_channel_cfg *channels_cfg;
+	uint32_t *const_values;
+	struct k_thread thread;
+	struct k_sem sem;
+
+	K_KERNEL_STACK_MEMBER(stack,
+			CONFIG_ADC_EMUL_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+/**
+ * @brief Standard ADC value function used when channel should return constant
+ *        value
+ *
+ * @param dev ADC emulator device
+ * @param channel Channel for which function is called
+ * @param param Pointer to mV value which should be returned
+ * @param result Pointer where result should be saved
+ *
+ * @return Always 0
+ */
+static int adc_emul_const_channel_value(const struct device *dev,
```

Comment:
I'm not very keen on using functions for this. Allowing a handler function to be provided by the test is one thing, but using one internally seems a step too far. The function is so trivial.

I think it would be better to have an enum that determines the behaviour, then use switch() on the enum when you need to get the value in the acquisition thread. It just seems simpler and easier to follow.

Reply:
I have added `enum adc_emul_input_source` to keep track of used input source. It is used in `adc_emul_get_chan_value()` function. New options to this enum will be added if there is need for new builtin input sources.

Reply:
Yes I like this a lot better, thanks